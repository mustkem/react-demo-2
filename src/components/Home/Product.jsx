import React, { useState, useEffect, useRef } from "react";
import { Link } from "react-router-dom";
import { LazyLoadImage } from "react-lazy-load-image-component";

import missingImage from "../../images/placeholder_for_missing_posters.png";

function Product({ item }) {
  const [imgSrc, setImgSrc] = useState("");
  const [askForImgFlag, setAskForImageFlag] = useState(true);
  const myRef = useRef(null);
  let imageLoadStatus = false;

  useEffect(() => {
    window.addEventListener("scroll", handleProductScroll);
    handleProductScroll();
    return () => {
      window.removeEventListener("scroll", handleProductScroll);
    };
  }, []);

  function lazyLoadImage() {
    try {
      const images = require.context("../../images", true);
      let img = images("./" + item["poster-image"]);
      setImgSrc(img);
    } catch {
      // console.log("image not available");
    }
  }

  function handleProductScroll() {
    if (imageLoadStatus) return false;

    var myElement = myRef.current;
    var bounding = myElement.getBoundingClientRect();

    if (
      bounding.top >= 0 &&
      bounding.left >= 0 &&
      bounding.right <= window.innerWidth &&
      bounding.top <= window.innerHeight
    ) {
      setAskForImageFlag(false);
      lazyLoadImage();
      imageLoadStatus = true;
      // console.log("Element is in the viewport!");
    } else {
      // console.log("Element is NOT in the viewport!");
    }
  }

  return (
    <li>
      <Link to="">
        {imgSrc && <img className="missing-img" src={imgSrc} alt="product" />}
        {!imgSrc && (
          <img
            ref={myRef}
            className="missing-img"
            src={missingImage}
            alt="product"
          />
        )}
        <span>{item.name}</span>
      </Link>
    </li>
  );
}

export default Product;
