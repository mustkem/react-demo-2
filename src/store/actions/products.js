import { path } from "ramda";

import httpInstance from "../../helpers/http-client";
import { actionTypes } from "../action-types/action-types";

export const onGetProducts = (data) => {
  return {
    type: actionTypes.GET_PRODUCTS,
    payload: data,
  };
};

export const getProducts = (page) => (dispatch, getState) => {
  return new Promise((res, rej) => {
    // dispatch(onGetProducts({ loading: true }));
    console.log("testtttt");
    httpInstance({
      method: "get",
      url: "/CONTENTLISTINGPAGE-PAGE" + page + ".json",
      responseType: "json", // default
    })
      .then(function (response) {
        const oldContent = path(
          ["products", "products", "page", "content-items", "content"],
          getState()
        );

        if (oldContent) {
          const responseData = response.data;
          responseData.page["content-items"].content = [
            ...oldContent,
            ...responseData.page["content-items"].content,
          ];
          dispatch(onGetProducts({ ...responseData, loading: false }));
        } else {
          dispatch(onGetProducts({ ...response.data, loading: false }));
        }
        res(response);
      })
      .catch(function (error) {
        console.log(error);
        rej();
      });
  });
};
