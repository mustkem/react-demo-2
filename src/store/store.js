import { createStore, applyMiddleware, compose } from "redux";
import rootReducer from "./reducers/rootReducers";
import thunk from "redux-thunk";

const composeEnhancers =
  (typeof window !== "undefined" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

export const appStore = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk))
);
