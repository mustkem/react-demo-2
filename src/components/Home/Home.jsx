import React from "react";
import Product from "./Product";
import { FiSearch } from "react-icons/fi";
import { BsArrowLeftShort } from "react-icons/bs";
import { MdClose } from "react-icons/md";
import { Link } from "react-router-dom";

class Home extends React.Component {
  constructor(props) {
    super(props);
    this.inputRef = React.createRef();

    this.state = {
      page: 1,
      searchString: "",
    };
  }

  componentDidMount() {
    this.props.getProducts(this.state.page);

    window.addEventListener("scroll", this.handleScroll);
  }

  handleScroll = () => {
    if (window.innerHeight + window.scrollY >= document.body.offsetHeight) {
      this.setState({
        page: this.state.page + 1,
      });
    }
  };

  componentDidUpdate(prevProp, prevState) {
    if (
      this.state.page !== prevState.page &&
      !this.props.loading &&
      prevState.page <= 3
    ) {
      this.props.getProducts(this.state.page);
    }
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.handleScroll);
  }

  filterItems = () => {
    return (
      this.props.products &&
      this.props.products.filter((item) => {
        const lowerCaseName = item.name.toLowerCase();
        const lowerCaseSearchString = this.state.searchString.toLowerCase();
        console.log();
        return lowerCaseName.includes(lowerCaseSearchString);
      })
    );
  };

  handleSearchBarModal = () => {
    this.setState(
      {
        searchBarModal: true,
      },
      () => {
        this.inputRef.current.focus();
      }
    );
  };

  render() {
    return (
      <div className="home-content main-content">
        <div className="top-sec">
          <div className="container">
            {!this.state.searchBarModal && (
              <div className="info">
                <Link to="">
                  <BsArrowLeftShort className="icon left-arrow-icon" />
                  {this.props.title}
                </Link>
                <FiSearch
                  onClick={this.handleSearchBarModal}
                  className="icon"
                />
              </div>
            )}
            {this.state.searchBarModal && (
              <div className="search-bar">
                <BsArrowLeftShort
                  onClick={() => this.setState({ searchBarModal: false })}
                  className="icon left-arrow-icon"
                />
                <input
                  ref={this.inputRef}
                  placeholder="Search"
                  className="search-input"
                  value={this.state.searchString}
                  onChange={(e) => {
                    this.setState({ searchString: e.target.value });
                  }}
                />
                <MdClose
                  onClick={() => this.setState({ searchString: "" })}
                  className="icon close"
                />
              </div>
            )}
          </div>
        </div>
        <div className="container container-products">
          <ul className="products">
            {this.filterItems() &&
              this.filterItems().map((item, index) => {
                return (
                  <Product
                    key={index + "" + this.props.pageNumRequested}
                    item={item}
                  />
                );
              })}
          </ul>
          {this.filterItems() && this.filterItems().length === 0 && (
            <div className="no-records">No Records </div>
          )}
        </div>
      </div>
    );
  }
}

export default Home;
