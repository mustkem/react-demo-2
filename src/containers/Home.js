import { connect } from "react-redux";
import { path } from "ramda";

import Home from "../components/Home/Home";

import { getProducts } from "../store/actions";

const mapStateToProps = (state) => {
  return {
    products: path(
      ["products", "products", "page", "content-items", "content"],
      state
    ),
    title: path(["products", "products", "page", "title"], state),
    pageNumRequested: path(
      ["products", "products", "page", "page-num-requested"],
      state
    ),
    loading: path(["products", "products", "loading"], state),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    getProducts: (index) => dispatch(getProducts(index)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
