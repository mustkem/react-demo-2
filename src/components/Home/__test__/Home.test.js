import React from "react";
import { shallow, mount } from "enzyme";
import Home from "../Home";
import { getProducts } from "../../../store/actions";

const props = {
  getProducts,
};

// snapsot testing //
it("Should render home page correctly", () => {
  const wrapper = shallow(<Home {...props} />);
  expect(wrapper).toMatchSnapshot();
});

// mock calling of api
it("Should calls call componentDidMount and thereafter call action.", () => {
  jest.spyOn(Home.prototype, "componentDidMount");
  const wrapper = shallow(<Home {...props} />);
  expect(Home.prototype.componentDidMount.mock.calls.length).toBe(1);
});

// Check if component accept props.
it("Should accept action props", () => {
  const wrapper = shallow(<Home {...props} />);
  expect(wrapper.instance().props.getProducts).toEqual(getProducts);
});
