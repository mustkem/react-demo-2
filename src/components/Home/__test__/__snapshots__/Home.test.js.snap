// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page render correctly 1`] = `ShallowWrapper {}`;

exports[`Should render home page correctly 1`] = `ShallowWrapper {}`;
